File: src/aiperf/timing/session_tree.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Per-session-TREE liveness bookkeeping for DAG-lineage finality.

A DAG session is not a single root conversation -- it is a whole TREE: a
depth-0 root plus every child it spawns, recursively (FORK/SPAWN children and
their subchildren). ``SessionTreeRegistry`` tracks the liveness of each such
tree, keyed by the tree's ``root_correlation_id`` (the depth-0 root's
x_correlation_id), so the credit issuer can stamp two conservative facts on
every emitted ``Credit`` at issue time:

  - ``is_parent_final`` -- the credit's parent had already returned its terminal
    turn (v1: determinable only when the parent IS the root).
  - ``is_tree_final`` -- this credit is provably the last request the whole tree
    will ever send.

This registry is **finality bookkeeping only**. Unlike the agentic-replay
variant it was ported from, it does NOT own session-slot release: on main the
credit callback handler releases the root's session slot on the root's terminal
turn (``_release_slots_for_return``) exactly as before, and DAG children inherit
the root's slot. The registry never touches the concurrency manager and fires no
drain callback -- it only answers the two finality queries below from live tree
state.

Wiring (DAG datasets only; the registry is None on non-DAG paths, so
``_finality_for_issue`` returns the conservative ``(None, False)``):
  - ``CreditIssuer._open_session_tree`` -> :meth:`open_tree` when a root
    session's first credit is issued (turn 0 session-slot acquisition).
  - ``BranchOrchestrator`` -> :meth:`register_descendants` when a parent turn
    spawns children, :meth:`on_descendant_done` at every descendant-terminal
    point, and :meth:`on_root_terminal` from ``intercept`` after the root's
    final-turn return is processed (so final-turn spawns register first).
  - ``BranchOrchestrator.cleanup`` -> :meth:`release_all` at phase teardown to
    drop any still-open trees.
"""

from __future__ import annotations

from dataclasses import dataclass

from aiperf.common.aiperf_logger import AIPerfLogger
from aiperf.common.enums import CreditPhase

_logger = AIPerfLogger(__name__)


@dataclass(slots=True)
class _TreeState:
    """Liveness of one session tree."""

    phase: CreditPhase
    """Phase the tree was opened under; teardown targets the same phase."""
    root_pending: bool
    """True until the root's terminal turn returns."""
    outstanding: int = 0
    """Descendants currently live or registered-pending (any depth)."""
    released: bool = False
    """Set once the tree has been retired; guards against double retire."""

    @property
    def drained(self) -> bool:
        """True when the root is done and no descendant work remains."""
        return not self.root_pending and self.outstanding <= 0


class SessionTreeRegistry:
    """Tracks per-session-tree liveness for DAG-lineage finality queries.

    See the module docstring for the invariant and wiring. Single-threaded
    asyncio: every method runs to completion between awaits, so the counter
    mutations are atomic without locking. Finality bookkeeping only -- this
    class never acquires or releases a concurrency slot.
    """

    def __init__(self) -> None:
        self._trees: dict[str, _TreeState] = {}
        # Descendants registered before their tree is opened. On main's flow a
        # root's tree is opened at turn-0 issue, before any child can spawn (a
        # spawn only happens on a parent turn RETURN), so this buffer is a
        # defensive belt that keeps the branch's state transitions identical.
        self._pending_descendants: dict[str, int] = {}
        # Roots whose tree already RETIRED (drained), mapped to the phase they
        # were opened under. A descendant's final-turn SPAWN registers its
        # grandchildren AFTER that descendant's own on_descendant_done drained
        # the tree (callback order: the child-completion decrement precedes the
        # return-intercept that spawns). Without this ledger those grandchildren
        # would buffer into a retired root that nothing drains, and is_tree_final
        # could never answer True for them. register_descendants consults it to
        # RESURRECT the tree root-terminal instead. Cleared at teardown.
        self._retired_roots: dict[str, CreditPhase] = {}
        # Peak simultaneously-open trees == peak tree concurrency; logged at
        # teardown so any overshoot is visible.
        self._peak_open: int = 0
        # A non-zero count means a descendant returned for a tree that was
        # already retired -- i.e. the tree drained while that work was still in
        # flight (premature drain). Surfaces span-overhang.
        self._late_events: int = 0

    def open_tree(
        self, root_corr: str, phase: CreditPhase, *, root_pending: bool
    ) -> None:
        """Record a newly admitted tree when its root's first credit is issued.

        Args:
            root_corr: the tree's ``root_correlation_id`` (depth-0 root's
                x_correlation_id).
            phase: phase the tree was opened under.
            root_pending: True when a root credit is in flight that will reach a
                terminal turn (always True on main -- open_tree fires only at a
                root session start).
        """
        existing = self._trees.get(root_corr)
        if existing is not None and not existing.released:
            # Duplicate open for a still-live tree: keep the original (its
            # outstanding count is authoritative). Should not happen with
            # correct wiring; log so a regression is visible.
            _logger.warning(
                lambda: f"open_tree for already-open tree root_corr={root_corr!r}; "
                "ignoring duplicate"
            )
            return
        state = _TreeState(phase=phase, root_pending=root_pending)
        state.outstanding += self._pending_descendants.pop(root_corr, 0)
        # A freshly-opened tree supersedes any stale retired record for the same
        # id (correlation ids are unique, so this is defensive).
        self._retired_roots.pop(root_corr, None)
        self._trees[root_corr] = state
        if len(self._trees) > self._peak_open:
            self._peak_open = len(self._trees)

    @property
    def peak_open(self) -> int:
        """Maximum simultaneously-open trees seen (== peak tree concurrency)."""
        return self._peak_open

    @property
    def late_events(self) -> int:
        """Count of returns for already-retired trees (premature-drain evidence)."""
        return self._late_events

    def has_tree(self, root_corr: str) -> bool:
        """True when this registry is tracking ``root_corr`` (engagement gate)."""
        return root_corr in self._trees

    def root_terminal(self, root_correlation_id: str) -> bool | None:
        """True when the tree's root has returned its terminal turn; None if unknown.

        Queried at credit-issue time while the tree is still live (a descendant
        being issued keeps it in ``_trees``). A fully drained tree has been
        retired and reads as unknown/None.
        """
        state = self._trees.get(root_correlation_id)
        if state is None:
            return None
        return not state.root_pending

    def is_last_tree_request(
        self,
        root_correlation_id: str,
        *,
        is_final_turn: bool,
        is_root_credit: bool,
        has_branches: bool,
    ) -> bool:
        """Conservative: True only when this credit is provably the tree's last request.

        ``has_branches`` means "this turn declares ANY branch (FORK or SPAWN)
        and will therefore spawn descendants on its return". It must NOT be the
        FORK-only ``has_forks`` flag: SPAWN children register with this registry
        only at return-intercept, AFTER the issuer stamped finality, so a
        spawning final turn would otherwise read as last while its children are
        pending (wrong-True, violating the conservative contract).
        """
        if not is_final_turn or has_branches:
            return False
        state = self._trees.get(root_correlation_id)
        if state is None:
            return False
        if is_root_credit:
            return state.outstanding <= 0
        return (not state.root_pending) and state.outstanding == 1

    def register_descendants(self, root_corr: str, n: int = 1) -> None:
        """Add ``n`` descendants (spawned under one root) to a tree.

        Called when a parent turn spawns children, keyed by the tree's root id.
        Three cases when the tree is not live:
          - RETIRED root (in ``_retired_roots``): resurrect it root-terminal.
            A descendant's final-turn SPAWN registers its grandchildren after
            that descendant's own ``on_descendant_done`` drained the tree; the
            root was terminal at retire (``drained`` requires ``not
            root_pending``), so recreate the tree with ``root_pending=False``
            and these descendants outstanding -- keeping ``is_tree_final``
            answerable and the count coherent (the grandchildren re-drain it
            when they finish).
          - Unopened root: buffer the count so a later ``open_tree`` folds it in
            (defensive -- on main the tree is always open first).
        """
        if n <= 0:
            return
        state = self._trees.get(root_corr)
        if state is not None:
            state.outstanding += n
            return
        retired_phase = self._retired_roots.pop(root_corr, None)
        if retired_phase is not None:
            self._trees[root_corr] = _TreeState(
                phase=retired_phase, root_pending=False, outstanding=n
            )
            if len(self._trees) > self._peak_open:
                self._peak_open = len(self._trees)
            return
        self._pending_descendants[root_corr] = (
            self._pending_descendants.get(root_corr, 0) + n
        )

    def on_descendant_done(self, root_corr: str) -> bool:
        """Account one descendant terminally completing (leaf / error / stopped /
        dispatch-rollback). Retires the tree iff it is now drained.

        Returns True if the tree was retired by this call.
        """
        state = self._trees.get(root_corr)
        if state is None:
            pending = self._pending_descendants.get(root_corr, 0)
            if pending > 0:
                self._pending_descendants[root_corr] = pending - 1
            else:
                self._late_events += 1
            return False
        if state.outstanding > 0:
            state.outstanding -= 1
        return self._retire_if_drained(root_corr, state)

    def on_root_terminal(self, root_corr: str) -> bool:
        """Account the root's terminal turn returning (or root cancel/truncate).

        Clears ``root_pending``; retires the tree iff it is now drained. Must be
        called AFTER the returning root credit's intercept has run, so any
        children spawned on the final turn are already registered.

        Returns True if the tree was retired by this call. False both when the
        tree is held (descendants remain) AND when ``root_corr`` is untracked.
        """
        state = self._trees.get(root_corr)
        if state is None:
            return False
        state.root_pending = False
        return self._retire_if_drained(root_corr, state)

    def _retire_if_drained(self, root_corr: str, state: _TreeState) -> bool:
        """Pop a drained tree from tracking. Finality-only: releases no slot."""
        if state.released or not state.drained:
            return False
        state.released = True
        self._trees.pop(root_corr, None)
        # Remember the retired root (with its phase) so a late final-turn SPAWN
        # can RESURRECT it instead of silently buffering. See register_descendants.
        self._retired_roots[root_corr] = state.phase
        return True

    def release_all(self, phase: CreditPhase | None = None) -> int:
        """Retire every still-open tree at teardown (optionally one phase).

        Finality-only: drops the tracking entries; releases no concurrency slot
        (the callback handler owns slot release on main). The registry is
        created per-phase, so ``phase=None`` (retire all) is the common
        teardown call. Returns the number of trees retired.
        """
        to_release = [
            root_corr
            for root_corr, state in self._trees.items()
            if (phase is None or state.phase == phase) and not state.released
        ]
        for root_corr in to_release:
            state = self._trees.pop(root_corr, None)
            if state is None or state.released:
                continue
            state.released = True
        # Drop the transient buffers: with the trees above retired, any
        # pending/retired descendant accounting refers to no live tree. On a
        # full teardown (``phase is None`` -- the common per-phase call, since
        # the registry is created per-phase) clear both; a phase-scoped call
        # clears only that phase's retired roots (``_pending_descendants``
        # carries no phase and is left untouched then).
        if phase is None:
            self._pending_descendants.clear()
            self._retired_roots.clear()
        else:
            self._retired_roots = {
                r: p for r, p in self._retired_roots.items() if p != phase
            }
        return len(to_release)

    def open_count(self, phase: CreditPhase | None = None) -> int:
        """Number of trees currently tracked (optionally for one phase)."""
        if phase is None:
            return len(self._trees)
        return sum(1 for state in self._trees.values() if state.phase == phase)

File: rust/runtime/src/agentx/session_tree.rs
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Per-session-tree slot-release accounting, ported from
//! `src/aiperf/timing/session_tree.py`.
//!
//! One session slot is held per live **tree** (a depth-0 root plus all recursive
//! descendants: subagents, flat-async, spawn/fork children), keyed by
//! `root_correlation_id`, and released exactly once when the tree DRAINS (root
//! done AND no descendant work outstanding). This bounds concurrency to the
//! configured slot count regardless of fan-out depth.
//!
//! The concurrency manager's `release_session_slot(phase)` is abstracted behind
//! [`SlotReleaser`]; the recycle `on_drain` hook is an optional callback. Phase
//! runtime keys are modeled as opaque string identities.

use std::collections::HashMap;

/// Opaque phase runtime key (the slot is released under the key it was acquired).
pub type PhaseKey = String;

/// The concurrency manager's slot-release seam.
pub trait SlotReleaser {
    /// Release one session slot acquired under `phase`.
    fn release_session_slot(&mut self, phase: &PhaseKey);
}

/// Liveness of one session tree (Python `_TreeState`).
#[derive(Debug, Clone)]
struct TreeState {
    phase: PhaseKey,
    root_pending: bool,
    outstanding: i64,
    released: bool,
}

impl TreeState {
    fn drained(&self) -> bool {
        !self.root_pending && self.outstanding <= 0
    }
}

/// Owns session-slot release per session tree (Python `SessionTreeRegistry`).
pub struct SessionTreeRegistry<R: SlotReleaser> {
    releaser: R,
    trees: HashMap<String, TreeState>,
    on_drain: Option<Box<dyn FnMut(&str, &PhaseKey)>>,
    pending_descendants: HashMap<String, i64>,
    peak_open: usize,
    late_events: i64,
}

impl<R: SlotReleaser> SessionTreeRegistry<R> {
    /// Construct with the concurrency-manager releaser.
    pub fn new(releaser: R) -> Self {
        Self {
            releaser,
            trees: HashMap::new(),
            on_drain: None,
            pending_descendants: HashMap::new(),
            peak_open: 0,
            late_events: 0,
        }
    }

    /// Register the drain callback (fired with `root_corr, phase` on normal
    /// release, NOT on `release_all` teardown).
    pub fn set_drain_callback(&mut self, callback: Option<Box<dyn FnMut(&str, &PhaseKey)>>) {
        self.on_drain = callback;
    }

    /// Maximum simultaneously-open trees seen (== peak session-slot occupancy).
    pub fn peak_open(&self) -> usize {
        self.peak_open
    }

    /// Count of returns for already-released trees (premature-drain evidence).
    pub fn late_events(&self) -> i64 {
        self.late_events
    }

    /// True when this registry is tracking `root_corr`.
    pub fn has_tree(&self, root_corr: &str) -> bool {
        self.trees.contains_key(root_corr)
    }

    /// Record a newly admitted tree after its session slot was acquired.
    pub fn open_tree(&mut self, root_corr: &str, phase: PhaseKey, root_pending: bool) {
        if let Some(existing) = self.trees.get(root_corr)
            && !existing.released
        {
            // Duplicate open for a still-live tree: keep the original.
            return;
        }
        let mut state = TreeState {
            phase,
            root_pending,
            outstanding: 0,
            released: false,
        };
        state.outstanding += self.pending_descendants.remove(root_corr).unwrap_or(0);
        self.trees.insert(root_corr.to_string(), state);
        if self.trees.len() > self.peak_open {
            self.peak_open = self.trees.len();
        }
    }

    /// Add `n` descendants (spawned or snapshot-seeded) to a tree. Buffered when
    /// the tree is not yet open.
    pub fn register_descendants(&mut self, root_corr: &str, n: i64) {
        if n <= 0 {
            return;
        }
        match self.trees.get_mut(root_corr) {
            Some(state) => state.outstanding += n,
            None => {
                *self
                    .pending_descendants
                    .entry(root_corr.to_string())
                    .or_insert(0) += n;
            }
        }
    }

    /// Account one descendant terminally completing. Releases the slot iff the
    /// tree is now drained. Returns true if the slot was released.
    pub fn on_descendant_done(&mut self, root_corr: &str) -> bool {
        if !self.trees.contains_key(root_corr) {
            // Tree not open yet: decrement the pending buffer, or count a late event.
            let pending = self
                .pending_descendants
                .get(root_corr)
                .copied()
                .unwrap_or(0);
            if pending > 1 {
                self.pending_descendants
                    .insert(root_corr.to_string(), pending - 1);
            } else if pending == 1 {
                self.pending_descendants.remove(root_corr);
            } else {
                self.late_events += 1;
            }
            return false;
        }
        if let Some(state) = self.trees.get_mut(root_corr)
            && state.outstanding > 0
        {
            state.outstanding -= 1;
        }
        self.maybe_release(root_corr)
    }

    /// Account the root's terminal turn returning. Clears `root_pending`;
    /// releases the slot iff the tree is now drained.
    pub fn on_root_terminal(&mut self, root_corr: &str) -> bool {
        if let Some(state) = self.trees.get_mut(root_corr) {
            state.root_pending = false;
        } else {
            return false;
        }
        self.maybe_release(root_corr)
    }

    fn maybe_release(&mut self, root_corr: &str) -> bool {
        let should = match self.trees.get(root_corr) {
            Some(s) => !s.released && s.drained(),
            None => false,
        };
        if !should {
            return false;
        }
        let state = self.trees.remove(root_corr).unwrap();
        self.releaser.release_session_slot(&state.phase);
        if let Some(cb) = self.on_drain.as_mut() {
            cb(root_corr, &state.phase);
        }
        true
    }

    /// Release every still-open tree's slot for `phase` at teardown. Does NOT
    /// fire the drain callback. Returns the number of slots released.
    pub fn release_all(&mut self, phase: &PhaseKey) -> usize {
        let to_release: Vec<String> = self
            .trees
            .iter()
            .filter(|(_, s)| &s.phase == phase && !s.released)
            .map(|(k, _)| k.clone())
            .collect();
        for root_corr in &to_release {
            if let Some(state) = self.trees.remove(root_corr)
                && !state.released
            {
                self.releaser.release_session_slot(phase);
            }
        }
        to_release.len()
    }

    /// Number of trees currently holding a slot (optionally for one phase).
    pub fn open_count(&self, phase: Option<&PhaseKey>) -> usize {
        match phase {
            None => self.trees.len(),
            Some(p) => self.trees.values().filter(|s| &s.phase == p).count(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Default)]
    struct RecRel {
        released: Vec<PhaseKey>,
    }
    impl SlotReleaser for RecRel {
        fn release_session_slot(&mut self, phase: &PhaseKey) {
            self.released.push(phase.clone());
        }
    }

    #[test]
    fn tree_drains_on_root_and_descendants() {
        let mut reg = SessionTreeRegistry::new(RecRel::default());
        reg.open_tree("root", "p0".into(), true);
        reg.register_descendants("root", 2);
        // Root terminal but descendants remain -> not released.
        assert!(!reg.on_root_terminal("root"));
        assert!(!reg.on_descendant_done("root"));
        // Last descendant done -> drained -> released.
        assert!(reg.on_descendant_done("root"));
        assert_eq!(reg.releaser.released, vec!["p0".to_string()]);
        assert!(!reg.has_tree("root"));
    }

    #[test]
    fn pending_descendants_folded_at_open() {
        let mut reg = SessionTreeRegistry::new(RecRel::default());
        // Snapshot seeds a descendant before the tree opens.
        reg.register_descendants("root", 1);
        reg.open_tree("root", "p0".into(), false); // rootless lane
        // The pre-seeded descendant keeps the tree alive until it completes.
        assert!(reg.on_descendant_done("root"));
    }

    #[test]
    fn descendant_done_before_open_decrements_pending() {
        let mut reg = SessionTreeRegistry::new(RecRel::default());
        reg.register_descendants("root", 2);
        // One completes before the tree opens -> pending folds to 1.
        assert!(!reg.on_descendant_done("root"));
        reg.open_tree("root", "p0".into(), false);
        // Only one descendant remains outstanding.
        assert!(reg.on_descendant_done("root"));
        assert_eq!(reg.late_events(), 0);
    }
}
